### PR TITLE
refactor(query): improve performance of aggregate function distinct

### PR DESCRIPTION
**Pull request by Kikkon**
Signed-off-by: kikkon <nian920@outlook.com>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

in debug mode:
before
```
mysql> select count(distinct number) from numbers_mt(10000000);
+------------------------+
| count(DISTINCT number) |
+------------------------+
|               10000000 |
+------------------------+
1 row in set (1 min 1.10 sec)
Read 10000000 rows, 76.29 MiB in 61.064 sec., 163.76 thousand rows/sec., 1.25 MiB/sec.
```

after
```
mysql> select count(distinct number) from numbers_mt(10000000);
+------------------------+
| count(DISTINCT number) |
+------------------------+
|               10000000 |
+------------------------+
1 row in set (22.38 sec)
Read 10000000 rows, 76.29 MiB in 22.338 sec., 447.66 thousand rows/sec., 3.42 MiB/sec.
```

## Changelog
improve performance of aggregate function distinct
## Related Issues
Related https://github.com/datafuselabs/databend/issues/6240

### CODE REVIEW

File: src/common/functions/src/aggregates/aggregate_combinator_distinct.rs [hunk 1]
```diff
@@ -262,22 +136,105 @@ impl AggregateFunction for AggregateDistinctCombinator {
...
+                > {
+                    nested_name: nested_name.to_owned(),
+                    arguments,
+                    nested,
+                    name,
+                    _s: PhantomData,
+                    _state: PhantomData,
+                }));
+            }, {
+                panic!("GetKVActionReply");
```
Comment (sundy-li) [inline src/common/functions/src/aggregates/aggregate_combinator_distinct.rs:211]: what's this ?

File: src/common/functions/src/aggregates/aggregate_combinator_distinct.rs [hunk 2]
```diff
@@ -262,22 +136,105 @@ impl AggregateFunction for AggregateDistinctCombinator {
...
+            > {
+                nested_name: nested_name.to_owned(),
+                arguments,
+                nested,
+                name,
+                _s: PhantomData,
+                _state: PhantomData,
+            }));
+        }
+        if phid.is_floating() {}
```
Comment (sundy-li) [inline src/common/functions/src/aggregates/aggregate_combinator_distinct.rs:227]: did not cover float.

File: src/common/datavalues/src/macros.rs [hunk 3]
```diff
@@ -282,6 +282,29 @@ macro_rules! with_match_primitive_types_error {
     }};
 }
 
+#[macro_export]
+macro_rules! with_match_primitive_types_error_2 {(
```
Comment (sundy-li) [inline src/common/datavalues/src/macros.rs:286]: it's not used.

File: src/common/functions/src/aggregates/aggregate_distinct_state.rs [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    fn is_empty(&self) -> bool {
+        self.set.is_empty()
+    }
+
+    fn len(&self) -> usize {
+        self.set.len()
+    }
+
+    fn add(&mut self, columns: &[ColumnRef], row: usize) -> Result<()> {
+        let value = columns.get(0).map(|s| s.get(row)).unwrap();
```
Comment (sundy-li) [inline src/common/functions/src/aggregates/aggregate_distinct_state.rs:196]: Need downcast the `columns[0]` to `PrimitiveColumn<T>`.

File: src/common/functions/src/aggregates/aggregate_distinct_state.rs [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    }
+
+    fn batch_add(
+        &mut self,
+        columns: &[ColumnRef],
+        validity: Option<&Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        for row in 0..input_rows {
+            let value = columns.get(0).map(|s| s.get(row)).unwrap();
```
Comment (sundy-li) [inline src/common/functions/src/aggregates/aggregate_distinct_state.rs:208]: ditto

File: src/common/functions/src/aggregates/aggregate_distinct_state.rs (hunk 4 above)
Comment (sundy-li) [reply src/common/functions/src/aggregates/aggregate_distinct_state.rs:196]: `s.get(row)` is slow path.

File: src/common/functions/src/aggregates/aggregate_combinator_distinct.rs (hunk 2 above)
Comment (Kikkon) [reply src/common/functions/src/aggregates/aggregate_combinator_distinct.rs:227]: 😅 Sorry, I missed some commits when I was using cherry pick, I'll commit them later with the other fixes

File: src/common/functions/src/aggregates/aggregate_distinct_state.rs (hunk 4 above)
Comment (Kikkon) [reply src/common/functions/src/aggregates/aggregate_distinct_state.rs]: 🤔 I feel that using value_unchecked does not improve performance significantly

Comment (sundy-li) [reply src/common/functions/src/aggregates/aggregate_distinct_state.rs]: Yes, but if we downcast the `columns[0]` to `PrimitiveColumn<T>`, we can avoid the virtual call.

File: src/common/functions/src/aggregates/aggregate_distinct_state.rs [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    fn build_columns(&self, fields: &[DataField]) -> Result<Vec<ColumnRef>>;
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Hash, Clone)]
+pub struct DataGroupValues(Vec<DataGroupValue>);
+
+pub struct AggregateDistinctState {
+    set: HashSet<DataGroupValues, RandomState>,
+}
+pub struct AggregateDistinctIntegerState<T: PrimitiveType + std::hash::Hash + Eq> {
```
Comment (sundy-li) [inline src/common/functions/src/aggregates/aggregate_distinct_state.rs:54]: Can we combine `AggregateDistinctIntegerState ` and `AggregateDistinctFloatState `  into one struct:

```
pub struct AggregatePrimitiveState<T: PrimitiveType, E: From<T>> {
    set: HashSet<E, RandomState>,
    _t: PhantomData<T>,
}
```

File: src/common/functions/src/aggregates/aggregate_distinct_state.rs [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+    }
+
+    fn build_columns(&self, fields: &[DataField]) -> Result<Vec<ColumnRef>> {
+        let mut results = Vec::with_capacity(self.set.len());
+        self.set.iter().for_each(|group_values| {
+            results.push(*group_values);
+        });
+
+        let data_type = fields[0].data_type();
+        let columns = results
```
Comment (sundy-li) [inline src/common/functions/src/aggregates/aggregate_distinct_state.rs:313]: `PrimitiveColumn::from_iterator` is enough.

File: src/common/functions/src/aggregates/aggregate_distinct_state.rs [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    }
+
+    fn build_columns(&self, fields: &[DataField]) -> Result<Vec<ColumnRef>> {
+        let mut results = Vec::with_capacity(self.set.len());
+        self.set.iter().for_each(|group_values| {
+            results.push(group_values.clone());
+        });
+
+        let data_type = fields[0].data_type();
+        let columns = results
```
Comment (sundy-li) [inline src/common/functions/src/aggregates/aggregate_distinct_state.rs:387]: `StringColumn::from_iterator` is enough.

**Comment by sundy-li** [comment]
Update in release mode:

```
databend > select count(distinct number) from numbers(10000000);
+------------------------+
| count(DISTINCT number) |
+------------------------+
|               10000000 |
+------------------------+
1 row in set (8.171 sec)

databend > select count(distinct number) from numbers(10000000);
+------------------------+
| count(DISTINCT number) |
+------------------------+
|                     32 |
+------------------------+
1 row in set (0.272 sec)
```

**Review by sundy-li — APPROVED** [review]

**Review by Xuanwo — APPROVED** [review]

**Comment by Kikkon** [comment]
Thank you for your support！ @sundy-li

**Pull request merged**